### PR TITLE
[SFEQS-791] Send the signature requests on app IO by messages 

### PR DESCRIPTION
**Pull request by grausof**
#### List of Changes
- Sending a signature request on `Storage Queue` when it is ready to be signed (all documents have been uploaded).
- Trigger of the `SendSignatureRequestMessage` function from the `Storage Queue`
- Sending a message to the IO app via a _standard service._
- Update to latest `handler-kit` API

#### How Has This Been Tested?

The project works locally and it is possible to try every part of it (thanks to `azurite`).
A test standard service was used to send the messages.

#### Types of changes

- [ ] Chore (nothing changes by a user perspective)
- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)

#### Checklist:

- [x] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -38,6 +38,21 @@ And run the project with
 npm start
 ```
 
+### Message services integration
+
+To use a message sending service to the IO app it is also necessary to set the following parameters in `local.settings.json`
```
Comment (gunzip) [inline README.md]: Since these look like local development instructions, it's useless to point to the real IO API since the IO App used for local development talks to the fake backend set up locally using docker. I

Ideally this service (io-functions-sign) should be run inside io-mock (the IO backend simulator) when you want to test the integration with other services. By now, I'd remove these instructions here (these var will be setup inside docker-compose anyway).

Comment (gunzip) [reply README.md]: How did you test the integration?

File: README.md [hunk 2]
```diff
@@ -97,3 +113,19 @@ To accomplish this, simply add the following property to the personal `settings.
 ```
 
 Please care of this setting, as it exposes the editor to potential security issues with third-party extensions.
+
+### Storage integration
+
+A storage account must also be configured to use blob storage and queues in `local.settings.json`
```
Comment (gunzip) [inline README.md]: Instead of documenting environment variables inside the readme, it's better to have a .env file with inline comments (just place comments above the variables). That's the .env included by docker-compose.

File: SendSignatureRequestMessage/function.json [hunk 3]
```diff
@@ -0,0 +1,12 @@
...
+  "bindings": [
+    {
+      "name": "sendSignatureRequestMessage",
+      "type": "queueTrigger",
+      "direction": "in",
+      "queueName": "waiting-for-signature",
+      "connection": "StorageAccountConnectionString"
+    }
+  ],
+  "scriptFile": "../lib/infra/azure/functions/send-signature-request-message.js"
```
Comment (gunzip) [inline SendSignatureRequestMessage/function.json:11]: The path here should be `../lib/ui/functions/send-signature-request-message.js`

File: package.json [hunk 4]
```diff
@@ -11,7 +11,8 @@
     "prepare": "husky install",
     "func:ext-install": "func extensions install",
     "start": "func start",
-    "generate:issuer-api-models": "gen-api-models --api-spec ./openapi-issuer.yaml --out-dir ./src/ui/api-models"
+    "generate:issuer-api-models": "gen-api-models --api-spec ./openapi-issuer.yaml --out-dir ./src/ui/api-models",
+    "generate:message-api-models": "gen-api-models --api-spec https://raw.githubusercontent.com/pagopa/io-functions-services/master/openapi/index.yaml --out-dir ./src/ui/api-models"
```
Comment (gunzip) [inline package.json]: This is a little controversial: if you point to the master branch (instead of a particular release) you get non reproducible builds. I'd use a specific releae in the path here. 

What about it @lucacavallaro ?

Ideally every service should publish it's SDK (API client) and other projects should use that one (handled with package.json) instead of generating the client locally. There's already a pipeline ready to do this, but I don't know why it's not used yet cc @gquadrati :-)

File: src/app/use-cases/send-signature-request.ts [hunk 5]
```diff
@@ -0,0 +1,98 @@
...
+    );
+
+export const sendMessage =
+  (message: NewMessage, signerId: string) =>
+  (
+    getFiscalCodeBySignerId: GetFiscalCodeBySignerId
+  ): TE.TaskEither<Error, CreatedMessage> =>
+    pipe(
+      sequenceS(TE.ApplySeq)({
+        signer_cf: getFiscalCodeBySignerId(signerId),
```
Comment (gunzip) [inline src/app/use-cases/send-signature-request.ts]:
```suggestion
        signerFiscalCode: getFiscalCodeBySignerId(signerId),
```

File: src/app/use-cases/send-signature-request.ts [hunk 6]
```diff
@@ -0,0 +1,98 @@
...
+      TE.map(({ product }) => mockMakeMessage(signatureRequest, product))
+    );
+
+export const sendMessage =
+  (message: NewMessage, signerId: string) =>
+  (
+    getFiscalCodeBySignerId: GetFiscalCodeBySignerId
+  ): TE.TaskEither<Error, CreatedMessage> =>
+    pipe(
+      sequenceS(TE.ApplySeq)({
```
Comment (gunzip) [inline src/app/use-cases/send-signature-request.ts]: Why are you using sequence here with just one parameter? you can just use chain.

File: src/app/use-cases/send-signature-request.ts [hunk 7]
```diff
@@ -0,0 +1,98 @@
...
+  (sendSignatureRequestBody: SendSignatureRequestBody) =>
+    pipe(
+      getSignatureRequest(
+        sendSignatureRequestBody.signatureRequestId,
+        sendSignatureRequestBody.subscriptionId
+      ),
+      TE.chainW(
+        flow(
+          TE.fromOption(() => signatureRequestNotFoundError),
+          TE.chain((sr) =>
```
Comment (gunzip) [inline src/app/use-cases/send-signature-request.ts]:
```suggestion
          TE.chain((signatureRequest) =>
```

please use meaningful names

File: src/app/use-cases/send-signature-request.ts (hunk 6 above)
Comment (gunzip) [reply src/app/use-cases/send-signature-request.ts]: Anyway, I won't make sendMessage dependant on `GetFiscalCodeBySignerId`, just call this method outside it and provide the `fiscalCode` directly (I guess you don't need this method at all since you can just pipe `submitMessageForUser(signerFiscalCode)`)

File: src/app/use-cases/send-signature-request.ts [hunk 8]
```diff
@@ -0,0 +1,98 @@
...
+      signatureRequest.expiresAt ? signatureRequest.expiresAt : "never"
+    }\`\n- docs: \`${JSON.stringify(signatureRequest.documents)}\`\n `,
+  },
+});
+
+export const prepareMessage =
+  (getProduct: GetProduct) =>
+  (signatureRequest: SignatureRequest): TE.TaskEither<Error, NewMessage> =>
+    pipe(
+      sequenceS(TE.ApplySeq)({
```
Comment (gunzip) [inline src/app/use-cases/send-signature-request.ts]: Just use chain when you have only one parameter (instead of sequence)

File: src/app/use-cases/validate-document.ts [hunk 9]
```diff
@@ -68,5 +71,10 @@ export const makeValidateDocument =
         )
       ),
       TE.chain(upsertSignatureRequest),
+      TE.chain((sr) =>
```
Comment (gunzip) [inline src/app/use-cases/validate-document.ts]:
```suggestion
      TE.chain((signatureRequest) =>
```

File: src/infra/azure/functions/send-signature-request-message.ts [hunk 10]
```diff
@@ -0,0 +1,50 @@
+import * as t from "io-ts";
```
Comment (gunzip) [inline src/infra/azure/functions/send-signature-request-message.ts:1]: this file must be inside `src/lib/functions`

File: src/infra/azure/functions/send-signature-request-message.ts [hunk 11]
```diff
@@ -0,0 +1,50 @@
...
+);
+
+export const SendSignatureRequestBody = t.type({
+  signatureRequestId: SignatureRequestId,
+  subscriptionId: SubscriptionId,
+});
+export type SendSignatureRequestBody = t.TypeOf<
+  typeof SendSignatureRequestBody
+>;
+
```
Comment (gunzip) [inline src/infra/azure/functions/send-signature-request-message.ts:50]: may you put a comment on what this method is supposed to do?

File: src/infra/azure/functions/validate-document-upload.ts [hunk 12]
```diff
@@ -40,10 +45,27 @@ const isDocumentUploadedToBlobStorage = pipe(
   )
 );
 
+const enqueueRequestAwaitingSignature = (request: SignatureRequest) =>
```
Comment (gunzip) [inline src/infra/azure/functions/validate-document-upload.ts:52]: please add a comment stating whatt this method is supposed to do

File: src/infra/azure/functions/validate-document-upload.ts [hunk 13]
```diff
@@ -9,7 +9,10 @@ import { last } from "fp-ts/ReadonlyNonEmptyArray";
 import { split } from "fp-ts/string";
```
Comment (gunzip) [inline src/infra/azure/functions/validate-document-upload.ts:9]: all function handlers must go into `src/lib/functions`

File: src/infra/azure/storage/queue/client.ts [hunk 14]
```diff
@@ -0,0 +1,24 @@
+import { QueueServiceClient, QueueClient } from "@azure/storage-queue";
+
+import * as TE from "fp-ts/TaskEither";
+
+export const createQueueClient = (
```
Comment (gunzip) [inline src/infra/azure/storage/queue/client.ts]: clients must be singletons shared by all functions, so what you realy want here is to export the instance directly (not a factory method)

see the `database` method into `database.ts`

File: src/infra/azure/storage/queue/client.ts [hunk 15]
```diff
@@ -0,0 +1,24 @@
...
+) => {
+  const queueServiceClient =
+    QueueServiceClient.fromConnectionString(connectionString);
+  return queueServiceClient.getQueueClient(queueName);
+};
+
+export const enqueueMessage = (queueClient: QueueClient) => (message: string) =>
+  TE.tryCatch(
+    () => queueClient.sendMessage(Buffer.from(message).toString("base64")),
+    () => new Error("Message cannot be queued")
```
Comment (gunzip) [inline src/infra/azure/storage/queue/client.ts]:
```suggestion
    toError
```

if you swallows errors this way we cannot have any info on why the error occurred

File: src/infra/azure/storage/queue/signature-request.ts [hunk 16]
```diff
@@ -0,0 +1,23 @@
...
+import { pipe } from "fp-ts/function";
+
+import { SignatureRequest } from "../../../../signature-request/signature-request";
+import { enqueueMessage } from "./client";
+
+export type EnqueueSignatureRequest = (
+  request: SignatureRequest
+) => TE.TaskEither<Error, SignatureRequest>;
+
+export const enqueueSignatureRequest =
```
Comment (gunzip) [inline src/infra/azure/storage/queue/signature-request.ts]: since there's nothing specific to azure here, I guess this file should be inside the use-case directory (the business logic goes there), isn't it @lucacavallaro ?

File: src/infra/services/config.ts [hunk 17]
```diff
@@ -0,0 +1,39 @@
...
+import * as E from "fp-ts/Either";
+import * as RE from "fp-ts/ReaderEither";
+import { pipe, flow } from "fp-ts/function";
+
+import { sequenceS } from "fp-ts/lib/Apply";
+
+import { IntegerFromString } from "@pagopa/ts-commons/lib/numbers";
+import { readFromNodeEnv } from "../../app/env";
+
+export const ServiceConfig = t.type({
```
Comment (gunzip) [inline src/infra/services/config.ts]: I'd rename everything to something like 'ServiceClient' or (better) 'IOApiClient'  since `Service` is too generic.

File: src/infra/services/send-message.ts [hunk 18]
```diff
@@ -0,0 +1,59 @@
...
+      TE.chain(({ basePath, headers, timeoutFetch, path }) =>
+        pipe(
+          TE.tryCatch(
+            () =>
+              timeoutFetch(`${basePath.href}${path.href}`, {
+                body: JSON.stringify(body),
+                method: "POST",
+                headers,
+              })
+                .then((r) => {
```
Comment (gunzip) [inline src/infra/services/send-message.ts]: it's better to don't mix Promise code with TaskEithers, just use map / filter ops.

File: src/infra/services/send-message.ts [hunk 19]
```diff
@@ -0,0 +1,59 @@
...
+                  if (!is2xx(r)) {
+                    throw Error("Unexpected webhook response");
+                  }
+                  return r;
+                })
+                .then((r) => r.json()),
```
Comment (gunzip) [inline src/infra/services/send-message.ts]: ditto

File: src/infra/services/send-message.ts [hunk 20]
```diff
@@ -0,0 +1,59 @@
...
+            E.toError
+          )
+        )
+      ),
+      TE.chain(
+        flow(
+          CreatedMessage.decode,
+          E.mapLeft(() => new Error("Invalid response")),
```
Comment (gunzip) [inline src/infra/services/send-message.ts]: please don't swallow errors, use `readableReport` method to output meaningful decoding errors instead. probably this should be handled at top level (inside handler-kit?): if the bubbled error is ValidationErrors output them with a human readable format

what about it @lucacavallaro ?

File: src/infra/services/service.ts [hunk 21]
```diff
@@ -0,0 +1,47 @@
...
+} from "@pagopa/ts-commons/lib/fetch";
+
+import * as E from "fp-ts/lib/Either";
+
+import { UrlFromString } from "@pagopa/ts-commons/lib/url";
+import { pipe } from "fp-ts/lib/function";
+import { Millisecond } from "@pagopa/ts-commons/lib/units";
+import { config } from "../../app/config";
+
+export const basePath = pipe(
```
Comment (gunzip) [inline src/infra/services/service.ts]: what about using the generated API client instead?

File: src/infra/services/service.ts [hunk 22]
```diff
@@ -0,0 +1,47 @@
...
+  config,
+  E.chain((config) =>
+    pipe(
+      config.service.serviceBasePath,
+      UrlFromString.decode,
+      E.mapLeft(() => new Error("Invalid service parameters"))
```
Comment (gunzip) [inline src/infra/services/service.ts]: don't swallow decoding errors

File: src/ui/api-models/Activation.ts [hunk 23]
```diff
@@ -0,0 +1,31 @@
+/**
```
Comment (gunzip) [inline src/ui/api-models/Activation.ts]: I'm not sure these generated files should be committed. Otherwise, every time the IO Api changes, we'll commit unrelated changes when pushgin PRs. What about it @lucacavallaro ?

**Review by gunzip — DISMISSED** [review]

File: package.json (hunk 4 above)
Comment (gquadrati) [reply package.json]: It's true, we have it https://www.npmjs.com/package/@pagopa/io-functions-services-sdk
If you don't need features in preview, you can use it

Comment (gunzip) [reply package.json]: Just check fp-ts and io-ts required versions since they're a dependency of the SDK as well @grausof

Comment (grausof) [reply package.json]: @gquadrati where can I find the documentation?

Comment (gquadrati) [reply package.json]: @grausof we are a bit lacking on specs :D 
But this is basically a package containing the same client and models you're generating right now.

File: src/infra/azure/storage/queue/signature-request.ts (hunk 16 above)
Comment (grausof) [reply src/infra/azure/storage/queue/signature-request.ts]: It is here because it depends on the `QueueClient` which is relative to `azure`

File: src/infra/services/send-message.ts (hunk 20 above)
Comment (grausof) [reply src/infra/services/send-message.ts]: Sorry, here I forgot to use `validate`

Comment (grausof) [reply src/infra/services/send-message.ts]: Fixed with: 
```
TE.chainW(
        flow(validate(CreatedMessage, "Invalid response"), TE.fromEither)
      )
```

File: src/app/use-cases/send-signature-request.ts [hunk 24]
```diff
@@ -0,0 +1,79 @@
...
+                  signatureRequest.signerId,
+                  getFiscalCodeBySignerId,
+                  TE.chain(submitMessageForUser(message))
+                )
+              )
+            )
+          )
+        )
+      ),
+      TE.map(() => void 0)
```
Comment (gunzip) [inline src/app/use-cases/send-signature-request.ts]:
```suggestion
```

File: src/app/use-cases/send-signature-request.ts [hunk 25]
```diff
@@ -0,0 +1,79 @@
...
+export const prepareMessage =
+  (getProduct: GetProduct) =>
+  (signatureRequest: SignatureRequest): TE.TaskEither<Error, NewMessage> =>
+    pipe(
+      getProduct(signatureRequest.productId, signatureRequest.subscriptionId),
+      TE.chainW(TE.fromOption(() => productNotFoundError)),
+      TE.map((product) => mockMakeMessage(signatureRequest, product))
+    );
+
+export const sendSignatureRequest =
```
Comment (gunzip) [inline src/app/use-cases/send-signature-request.ts:48]: may you add a comment on what this method does?

File: src/app/use-cases/validate-document.ts [hunk 26]
```diff
@@ -68,5 +71,10 @@ export const makeValidateDocument =
         )
       ),
       TE.chain(upsertSignatureRequest),
+      TE.chain((signatureRequest) =>
+        status(signatureRequest) === "WAIT_FOR_SIGNATURE"
+          ? enqueueRequestAwaitingSignature(signatureRequest)
+          : TE.right(signatureRequest)
+      ),
```
Comment (gunzip) [inline src/app/use-cases/validate-document.ts]:
```suggestion
      TE.filterOrElse(flow(status,  !S.Eq.equals("WAIT_FOR_SIGNATURE")),
          enqueueRequestAwaitingSignature
      ),
```

File: src/app/use-cases/validate-document.ts [hunk 27]
```diff
@@ -68,5 +71,10 @@ export const makeValidateDocument =
...
       TE.map(() => void 0)
```
Comment (gunzip) [inline src/app/use-cases/validate-document.ts]:
```suggestion
```

File: src/app/use-cases/validate-document.ts [hunk 28]
```diff
@@ -68,5 +71,10 @@ export const makeValidateDocument =
         )
```
Comment (gunzip) [inline src/app/use-cases/validate-document.ts:71]: may you add a comment on what the method 'makeValidateDocument' is supposed to do? explaining why we're putting the message into the queue.

File: src/infra/azure/storage/queue/signature-request.ts (hunk 16 above)
Comment (gunzip) [reply src/infra/azure/storage/queue/signature-request.ts]: I see. It looks like the only thing this method add to `enqueueMessage` is the `JSON.stringify `part, probably you don't need this method at all as is. 

What you really want is to define a new type for each of the different kind of messages and an encoder (JSON.stringify) for them. In this case, it would be a method that takes a `signatureRequest` object and converts it to the structure the queue message is expecting. This would enable the consumer and producer to share this type to encode / decode it.

Something like: `SignatureRequestMessage.encode(signatureRequest)`

File: src/ui/api-models/Clause.ts [hunk 29]
```diff
@@ -10,7 +10,7 @@ import * as t from "io-ts";
 const ClauseR = t.interface({
   title: t.string,
 
-  required: t.boolean,
+  required: t.boolean
```
Comment (gunzip) [inline src/ui/api-models/Clause.ts]: I'm unsure if we want to commit generated files. cc @lucacavallaro what about it?

File: src/app/use-cases/validate-document.ts (hunk 26 above)
Comment (grausof) [reply src/app/use-cases/validate-document.ts]: `enqueueRequestAwaitingSignature` return a `TaskEither` and I can't use a `filterOrElse`, furthermore `S.Eq.equals` requires you to pass the second parameter without a pipe. 
I could change like that:

```
TE.chain((signatureRequest) =>
        S.Eq.equals("WAIT_FOR_SIGNATURE", status(signatureRequest))
          ? enqueueRequestAwaitingSignature(signatureRequest)
          : TE.right(signatureRequest)
      )
```

Comment (gunzip) [reply src/app/use-cases/validate-document.ts]: sorry, no it's ok as is

File: src/ui/api-models/Clause.ts (hunk 29 above)
Comment (grausof) [reply src/ui/api-models/Clause.ts]: for me it is better not to commit them.

Comment (gunzip) [reply src/ui/api-models/Clause.ts]: deal, put them into some `generated` folder then add them to .gitingore

File: src/app/use-cases/validate-document.ts (hunk 27 above)
Comment (grausof) [reply src/app/use-cases/validate-document.ts]: I note that `TE.map(() => void 0)` is necessary for the handler-kit to work properly. CC @lucacavallaro

File: src/ui/api-models/Clause.ts (hunk 29 above)
Comment (grausof) [reply src/ui/api-models/Clause.ts]: in this way we can delete the `ui` folder?

Comment (gunzip) [reply src/ui/api-models/Clause.ts]: Yes if it just contains generated files. 

Consider to generate the types for queue message (from a different yaml) as well. In other words, since other processes (external to this repo) may read the queue messages, better to threat them as API types.

Comment (grausof) [reply src/ui/api-models/Clause.ts]: ok I move everything to a `generated` folder. For now, the types for queue messages are used only by this repository. However as soon as possible we will create a dedicated yaml.

File: src/infra/services/service.ts (hunk 21 above)
Comment (grausof) [reply src/infra/services/service.ts]: Done!

File: local.settings.json.example [hunk 30]
```diff
@@ -0,0 +1,15 @@
+{
```
Comment (gunzip) [inline local.settings.json.example:1]: can we use a .env file instead? (so we can add commments)

File: src/infra/IOApiClient/client.ts [hunk 31]
```diff
@@ -0,0 +1,38 @@
...
+      fetchApi: toFetch(
+        setFetchTimeout(
+          config.ioapi.serviceRequestTimeout as Millisecond,
+          abortableFetch
+        )
+      ) as unknown as typeof fetch,
+      withDefaults: (op) => (params) =>
+        op({
+          ...params,
+          // please refer to source api spec for actual header mapping
```
Comment (gunzip) [inline src/infra/IOApiClient/client.ts]: is this comment relevant?

Comment (grausof) [reply src/infra/IOApiClient/client.ts]: We can take it off!

File: local.settings.json.example (hunk 30 above)
Comment (grausof) [reply local.settings.json.example:1]: `local.settings.json` is the default file used by azure function just for local usage but in production, these values have to be populated in Application settings. I think it is not possible to rename it!

Comment (lucacavallaro) [reply local.settings.json.example:1]: maybe we can add a `local.settings.json.example` file

Comment (grausof) [reply local.settings.json.example:1]: @lucacavallaro already added in PR!

**Review by lucacavallaro — APPROVED** [review]
Bravo Francesco 👏
